docs/iid-1.md

```diff
@@ -0,0 +1,297 @@
+
+The BoringCrypto library provides a C-language API for use by other processes that require cryptographic functionality. All operations of the module occur via calls from host applications and their respective internal daemons/processes. The cryptographic boundary of the BoringCrypto module is a single object file named bcm.o which is statically linked to BoringSSL. The library performs no communications other than with the calling application (the process that invokes the library services) and the host operating system.
+
+
+### Implementation
+
+Implementation proposal for **immudb** FIPS binary is by creating a new `Dockerfile.fips` which builds from the boringcrypto golang image. This helps in the following ways:
+
+- No changes required in current build
+- FIP compliant images are available only for amd64, and **not for arm64 builds**. A separate Dockerfile will help in creating a separate workflow stage during CI build to export the FIPS compliant binaries.
+- Separate stage in CI can be added to run tests from the FIPS compliant build
+
+Google [maintains](https://github.com/golang/go/blob/dev.boringcrypto.go1.18/misc/boring/README.md) docker images with patched go toolchain, similar to golang:x.y.z.
+```
+A Dockerfile that starts with FROM golang:1.17.2 can switch to FROM us-docker.pkg.dev/google.com/api-project-999119582588/go-boringcrypto/golang:1.17.2b7 and should need no other modifications.
+```
+
+Upstream golang/go repository maintains a separate dev branch dev.boringcrypto.go<go-version>. This branch holds all the patches to make Go using BoringCrypto and the community has specified that there is no intention to merge this branch to main. The [patches](https://github.com/golang/go/branches/all?query=dev.boringcrypto) are kept up-to-date against minor version releases and master.
+
+
+##### Building Immudb
+This section describes how the immudb container image can be compiled and linked to BoringCrypto for FIPS compliance.
+
+- The Dockerfile.fips can be used for building immudb from the goboring image for Go 1.18 (us-docker.pkg.dev/google.com/api-project-999119582588/go-boringcrypto/golang:1.18.5b7).
+
+- Enable cgo: `CGO_ENABLED=1` ensures `cgo` is enabled in the immudb compilation process
+
+- To ensure the fips compliant `immudb` binary is linked with the BoringCrypto C library, we can run the `go tool nm` command which lists the symbols defined or used by an object file, archive, or executable. The presence of boring crypto can be analysed by checking if the `_Cfunc__goboringcrypto_` string is present. To ensure FIPS-only schemes are used, a check on the symbols is peformed for `crypto/internal/boring/sig.FIPSOnly` string. This is part of a script `check-fips.sh` which can be run on FIPS docker build for verification.
+
+- With boringcrypto image, a new function is added to the crypto/tls pkg `crypto/tls/fipsonly` which can be used to ensure fips mode has been enabled.
```

Comment:
immudb should log when running in FIPS compliant mode e.g. when logging the version and commit hash